File: packages/python/elizaos/advanced_capabilities/personality/services/character_file_manager.py
"""CharacterFileManager -- safe character file modifications with backup and validation.

Handles backup, validation, and atomic updates of character definition files.
Operates in memory-only mode when no character file is detected on disk.
"""

from __future__ import annotations

import json
import logging
import os
import re
import shutil
import time
from pathlib import Path
from typing import TYPE_CHECKING, Any

from google.protobuf.json_format import MessageToDict, ParseDict

from elizaos.types import MemoryType, Service
from elizaos.types.agent import CharacterSettings, StyleGuides

from ..types import PERSONALITY_SERVICE_TYPE

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Validation schema (mirrors the Zod schema from TypeScript)
# ---------------------------------------------------------------------------

_NAME_RE = re.compile(r"^[a-zA-Z0-9\s\-_]+$")
_FORBIDDEN_NAME_WORDS = {"admin", "system", "root"}
_INJECTION_PHRASES = {
    "ignore previous instructions",
    "disregard",
    "forget everything",
}


def _validate_name(value: Any) -> bool:
    if not isinstance(value, str):
        return False
    name = value.strip()
    return (
        0 < len(name) < 100
        and bool(_NAME_RE.match(name))
        and not any(word in name.lower() for word in _FORBIDDEN_NAME_WORDS)
        and name == name.strip()
    )


def _validate_system(value: Any) -> bool:
    if not isinstance(value, str):
        return False
    s = value
    return (
        10 < len(s) < 10_000
        and "<script>" not in s
        and "javascript:" not in s
        and "eval(" not in s
        and "Function(" not in s
        and not any(phrase in s.lower() for phrase in _INJECTION_PHRASES)
    )


def _validate_bio(value: Any) -> bool:
    if not isinstance(value, list):
        return False
    return all(
        isinstance(item, str)
        and 0 < len(item) < 500
        and "<script>" not in item
        and "javascript:" not in item
        for item in value
    )


def _validate_topics(value: Any) -> bool:
    if not isinstance(value, list):
        return False
    return all(isinstance(t, str) and 0 < len(t) < 100 and bool(_NAME_RE.match(t)) for t in value)


_FIELD_VALIDATORS: dict[str, Any] = {
    "name": _validate_name,
    "system": _validate_system,
    "bio": _validate_bio,
    "topics": _validate_topics,
}


class CharacterFileManager(Service):
    """Service for safely managing character file modifications."""

    name = "character_file_manager"
    service_type = PERSONALITY_SERVICE_TYPE

    @property
    def capability_description(self) -> str:
        return "Manages safe character file modifications with backup and validation"

    def __init__(self, runtime: IAgentRuntime | None = None) -> None:
        self._runtime: IAgentRuntime | None = runtime
        self._character_file_path: str | None = None
        self._backup_dir: str = os.path.join(os.getcwd(), ".eliza", "character-backups")
        self._max_backups: int = 10

    @classmethod
    async def start(cls, runtime: IAgentRuntime) -> CharacterFileManager:
        manager = cls(runtime)
        await manager._initialize()
        return manager

    async def _initialize(self) -> None:
        Path(self._backup_dir).mkdir(parents=True, exist_ok=True)
        await self._detect_character_file()
        logger.debug(
            "CharacterFileManager initialized: file=%s, backup=%s",
            self._character_file_path,
            self._backup_dir,
        )

    async def _detect_character_file(self) -> None:
        if not self._runtime:
            return

        character = self._runtime.character
        char_name = character.name if hasattr(character, "name") else ""

        cwd = os.getcwd()
        possible_paths = [
            os.path.join(cwd, f"{char_name}.json"),
            os.path.join(cwd, "character.json"),
            os.path.join(cwd, "agent", f"{char_name}.json"),
            os.path.join(cwd, "agent", "character.json"),
            os.path.join(cwd, "characters", f"{char_name}.json"),
            os.path.join(cwd, "characters", "character.json"),
        ]

        for file_path in possible_paths:
            if os.path.exists(file_path):
                try:
                    with open(file_path, encoding="utf-8") as f:
                        content = json.load(f)
                    if content.get("name") == char_name:
                        self._character_file_path = file_path
                        logger.debug("Character file detected: %s", file_path)
                        return
                except Exception:
                    continue

        logger.debug("No character file on disk, operating in memory-only mode")

    async def create_backup(self) -> str | None:
        """Create a backup of the current character file."""
        if not self._character_file_path:
            logger.warning("No character file path available for backup")
            return None

        try:
            timestamp = time.strftime("%Y%m%dT%H%M%S")
            basename = os.path.splitext(os.path.basename(self._character_file_path))[0]
            backup_name = f"{basename}-{timestamp}.json"
            backup_path = os.path.join(self._backup_dir, backup_name)

            shutil.copy2(self._character_file_path, backup_path)
            await self._cleanup_old_backups()
            logger.info("Character backup created: %s", backup_path)
            return backup_path
        except Exception as e:
            logger.error("Failed to create character backup: %s", e)
            return None

    async def _cleanup_old_backups(self) -> None:
        try:
            files = []
            for f in os.listdir(self._backup_dir):
                if f.endswith(".json"):
                    fpath = os.path.join(self._backup_dir, f)
                    files.append((fpath, os.path.getmtime(fpath)))
            files.sort(key=lambda x: x[1], reverse=True)

            for fpath, _ in files[self._max_backups :]:
                os.unlink(fpath)

            removed = max(0, len(files) - self._max_backups)
            if removed:
                logger.info("Cleaned up %d old backups", removed)
        except Exception as e:
            logger.error("Error cleaning up old backups: %s", e)

    def validate_modification(self, modification: dict[str, Any]) -> dict[str, Any]:
        """Validate a character modification. Returns {valid, errors}."""
        errors: list[str] = []

        # Field validation
        for field_name, validator in _FIELD_VALIDATORS.items():
            if field_name in modification:
                if not validator(modification[field_name]):
                    errors.append(f"Invalid {field_name}: failed validation rules")

        # Safety checks
        bio = modification.get("bio")
        if isinstance(bio, list) and len(bio) > 20:
            errors.append("Too many bio elements - maximum 20 allowed")

        topics = modification.get("topics")
        if isinstance(topics, list) and len(topics) > 50:
            errors.append("Too many topics - maximum 50 allowed")

        return {"valid": len(errors) == 0, "errors": errors}

    async def apply_modification(self, modification: dict[str, Any]) -> dict[str, Any]:
        """Apply a validated modification to the character."""
        validation = self.validate_modification(modification)
        if not validation["valid"]:
            return {
                "success": False,
                "error": f"Validation failed: {', '.join(validation['errors'])}",
            }

        if not self._runtime:
            return {"success": False, "error": "Runtime not available"}

        try:
            await self.create_backup()

            character = self._runtime.character
            getattr(character, "name", None)

            # Apply modifications (additive merge, not replacement)
            if "name" in modification:
                old_name = getattr(character, "name", "")
                character.name = modification["name"]
                logger.info("Character name changed: %s -> %s", old_name, modification["name"])

            if "system" in modification:
                character.system = modification["system"]
                logger.info("System prompt modified")

            if "bio" in modification:
                current_bio = getattr(character, "bio", [])
                if isinstance(current_bio, str):
                    current_bio = [current_bio]
                elif not isinstance(current_bio, list):
                    current_bio = []

                new_elements = [
                    b
                    for b in modification["bio"]
                    if not any(
                        existing.lower() in b.lower() or b.lower() in existing.lower()
                        for existing in current_bio
                    )
                ]
                character.bio = current_bio + new_elements

            if "topics" in modification:
                current_topics = getattr(character, "topics", []) or []
                new_topics = [t for t in modification["topics"] if t not in current_topics]
                character.topics = current_topics + new_topics

            if "message_examples" in modification:
                current_examples = getattr(character, "message_examples", []) or []
                character.message_examples = current_examples + modification["message_examples"]

            if "style" in modification:
                style_update = modification["style"]
                if isinstance(style_update, dict):
                    current_style = MessageToDict(character.style, preserving_proto_field_name=True)
                    for key, value in style_update.items():
                        existing = current_style.get(key, [])
                        existing_list = existing if isinstance(existing, list) else []
                        incoming_list = value if isinstance(value, list) else [value]
                        current_style[key] = existing_list + incoming_list
                    character.style.CopyFrom(ParseDict(current_style, StyleGuides()))

            if "settings" in modification:
                settings_update = modification["settings"]
                if isinstance(settings_update, dict):
                    current_settings = MessageToDict(
                        character.settings, preserving_proto_field_name=True
                    )
                    current_settings.update(settings_update)
                    character.settings.CopyFrom(ParseDict(current_settings, CharacterSettings()))

            # Write to file if available
            if self._character_file_path:
                char_dict = {}
                for attr in (
                    "name",
                    "system",
                    "bio",
                    "topics",
                    "message_examples",
                    "style",
                    "settings",
                ):
                    val = getattr(character, attr, None)
                    if val is not None:
                        char_dict[attr] = val
                with open(self._character_file_path, "w", encoding="utf-8") as f:
                    json.dump(char_dict, f, indent=2, default=str)
                logger.info("Character file updated successfully")

            # Audit log (best-effort)
            try:
                await self._runtime.create_memory(
                    {
                        "entityId": str(self._runtime.agent_id),
                        "roomId": str(self._runtime.agent_id),
                        "content": {
                            "text": f"Character modification applied to file: {self._character_file_path or 'memory-only'}",
                            "source": "character_modification",
                        },
                        "metadata": {
                            "type": MemoryType.CUSTOM,
                            "service": PERSONALITY_SERVICE_TYPE,
                            "action": "character_modified",
                            "timestamp": int(time.time() * 1000),
                            "filePath": self._character_file_path,
                            "modificationType": "file_update",
                        },
                    },
                    "character_modifications",
                )
            except Exception as e:
                logger.warning("Character modification audit log failed: %s", e)

            return {"success": True}
        except Exception as e:
            logger.error("Failed to apply character modification: %s", e)
            return {"success": False, "error": f"Application failed: {e}"}

    async def get_modification_history(self, limit: int = 10) -> list[dict[str, Any]]:
        """Get recent modification history."""
        if not self._runtime:
            return []

        memories = await self._runtime.get_memories(
            {
                "entityId": str(self._runtime.agent_id),
                "count": limit,
                "tableName": "character_modifications",
            }
        )

        results: list[dict[str, Any]] = []
        for memory in memories:
            meta = memory.metadata if hasattr(memory, "metadata") else {}
            if isinstance(meta, dict):
                results.append(
                    {
                        "timestamp": meta.get("timestamp"),
                        "modification": meta.get("modification"),
                        "filePath": meta.get("filePath"),
                    }
                )
        return results

    async def get_available_backups(self) -> list[dict[str, Any]]:
        """Get list of available backups."""
        if not os.path.exists(self._backup_dir):
            return []

        backups: list[dict[str, Any]] = []
        for filename in os.listdir(self._backup_dir):
            if filename.endswith(".json"):
                fpath = os.path.join(self._backup_dir, filename)
                stat = os.stat(fpath)
                backups.append(
                    {
                        "path": fpath,
                        "timestamp": int(stat.st_mtime * 1000),
                        "size": stat.st_size,
                    }
                )
        return sorted(backups, key=lambda b: b["timestamp"], reverse=True)

    async def restore_from_backup(self, backup_path: str) -> dict[str, Any]:
        """Restore character from a backup file."""
        if not os.path.exists(backup_path):
            return {"success": False, "error": "Backup file not found"}

        try:
            with open(backup_path, encoding="utf-8") as f:
                backup_content = json.load(f)

            if not isinstance(backup_content.get("name"), str):
                return {"success": False, "error": "Invalid backup file format"}

            # Create backup of current state
            await self.create_backup()

            # Update runtime character
            if self._runtime:
                for key, value in backup_content.items():
                    setattr(self._runtime.character, key, value)

            # Update file
            if self._character_file_path:
                with open(self._character_file_path, "w", encoding="utf-8") as f:
                    json.dump(backup_content, f, indent=2, default=str)

            logger.info("Character restored from backup: %s", backup_path)
            return {"success": True}
        except Exception as e:
            logger.error("Failed to restore from backup: %s", e)
            return {"success": False, "error": f"Restoration failed: {e}"}

    async def stop(self) -> None:
        logger.info("CharacterFileManager stopped")

File: packages/typescript/src/features/advanced-capabilities/personality/services/character-file-manager.ts
import path from "node:path";
import fs from "fs-extra";
import { z } from "zod";
import { logger } from "../../../../logger.ts";
import type {
	IAgentRuntime,
	MessageExample,
	MessageExampleGroup,
} from "../../../../types/index.ts";
import { MemoryType } from "../../../../types/memory.ts";
import { Service } from "../../../../types/service.ts";
import { PersonalityServiceType } from "../types.ts";

// Validation schema for character modifications
const CharacterModificationSchema = z.object({
	name: z.string().optional().describe("Character name"),
	system: z
		.string()
		.optional()
		.describe("System prompt that defines agent behavior and instructions"),
	bio: z.array(z.string()).optional(),
	messageExamples: z
		.array(
			z.array(
				z.object({
					name: z.string(),
					content: z.object({
						text: z.string(),
						actions: z.array(z.string()).optional(),
					}),
				}),
			),
		)
		.optional(),
	topics: z.array(z.string()).optional(),
	style: z
		.object({
			all: z.array(z.string()).optional(),
			chat: z.array(z.string()).optional(),
			post: z.array(z.string()).optional(),
		})
		.optional(),
	settings: z
		.record(
			z.string(),
			z.union([z.string(), z.number(), z.boolean(), z.null()]),
		)
		.optional(),
});

type CharacterModification = z.infer<typeof CharacterModificationSchema>;

type CharacterPersistenceServiceLike = {
	persistCharacter: (params?: {
		character?: Record<string, unknown>;
		previousName?: string;
	}) => Promise<{ success: boolean; error?: string }>;
};

/**
 * Service for safely managing character file modifications
 * Handles backup, validation, and atomic updates of character files
 */
export class CharacterFileManager extends Service {
	static serviceType = PersonalityServiceType.CHARACTER_MANAGEMENT;

	capabilityDescription =
		"Manages safe character file modifications with backup and validation";

	private characterFilePath: string | null = null;
	private backupDir: string;
	private maxBackups = 10;
	private validationRules: Map<string, (value: unknown) => boolean> = new Map();

	constructor(runtime?: IAgentRuntime) {
		super(runtime);
		this.backupDir = path.join(process.cwd(), ".eliza", "character-backups");
		this.setupValidationRules();
	}

	static async start(runtime: IAgentRuntime): Promise<CharacterFileManager> {
		const manager = new CharacterFileManager(runtime);
		await manager.initialize();
		return manager;
	}

	private async initialize(): Promise<void> {
		// Ensure backup directory exists
		await fs.ensureDir(this.backupDir);

		// Try to detect the character file path
		await this.detectCharacterFile();

		logger.debug(
			{ characterFile: this.characterFilePath, backupDir: this.backupDir },
			"CharacterFileManager initialized",
		);
	}

	private async detectCharacterFile(): Promise<void> {
		const character = this.runtime.character;

		// Look for character file in common locations
		const possiblePaths = [
			// Current working directory
			path.join(process.cwd(), `${character.name}.json`),
			path.join(process.cwd(), "character.json"),

			// Agent directory
			path.join(process.cwd(), "agent", `${character.name}.json`),
			path.join(process.cwd(), "agent", "character.json"),

			// Characters directory
			path.join(process.cwd(), "characters", `${character.name}.json`),
			path.join(process.cwd(), "characters", "character.json"),

			// Relative paths
			path.join(process.cwd(), "..", "characters", `${character.name}.json`),
			path.join(
				process.cwd(),
				"..",
				"..",
				"characters",
				`${character.name}.json`,
			),
		];

		for (const filePath of possiblePaths) {
			if (await fs.pathExists(filePath)) {
				try {
					const content = await fs.readJSON(filePath);
					if (content.name === character.name) {
						this.characterFilePath = filePath;
						logger.debug({ path: filePath }, "Character file detected");
						return;
					}
				} catch {
					// Continue searching
				}
			}
		}

		logger.debug("No character file on disk, operating in memory-only mode");
	}

	private setupValidationRules(): void {
		// Name validation - ensure safe and reasonable names
		this.validationRules.set("name", (value: unknown) => {
			if (typeof value !== "string") {
				return false;
			}
			const name = value;
			return (
				name.length > 0 &&
				name.length < 100 &&
				/^[a-zA-Z0-9\s\-_]+$/.test(name) && // Alphanumeric, spaces, hyphens, underscores only
				!name.toLowerCase().includes("admin") && // Prevent impersonation
				!name.toLowerCase().includes("system") &&
				!name.toLowerCase().includes("root") &&
				!name.trim().startsWith(" ") && // No leading/trailing spaces
				!name.trim().endsWith(" ")
			);
		});

		// System prompt validation - ensure safe and reasonable content
		this.validationRules.set("system", (value: unknown) => {
			if (typeof value !== "string") {
				return false;
			}
			const system = value;
			return (
				system.length > 10 && // Minimum meaningful length
				system.length < 10000 && // Maximum reasonable length
				!system.includes("<script>") && // Basic XSS protection
				!system.includes("javascript:") &&
				!system.includes("eval(") &&
				!system.includes("Function(") &&
				!system.toLowerCase().includes("ignore previous instructions") && // Prompt injection protection
				!system.toLowerCase().includes("disregard") &&
				!system.toLowerCase().includes("forget everything")
			);
		});

		// Bio validation - ensure reasonable length and content
		this.validationRules.set("bio", (value: unknown) => {
			if (!Array.isArray(value)) {
				return false;
			}
			return value.every(
				(item) =>
					typeof item === "string" &&
					item.length > 0 &&
					item.length < 500 &&
					!item.includes("<script>") && // Basic XSS protection
					!item.includes("javascript:"),
			);
		});

		// Topics validation
		this.validationRules.set("topics", (value: unknown) => {
			if (!Array.isArray(value)) {
				return false;
			}
			return value.every(
				(topic) =>
					typeof topic === "string" &&
					topic.length > 0 &&
					topic.length < 100 &&
					/^[a-zA-Z0-9\s\-_]+$/.test(topic), // Alphanumeric, spaces, hyphens, underscores only
			);
		});
	}

	async createBackup(): Promise<string | null> {
		if (!this.characterFilePath) {
			logger.warn("No character file path available for backup");
			return null;
		}

		try {
			const timestamp = new Date().toISOString().replace(/[:.]/g, "-");
			const backupFileName = `${path.basename(this.characterFilePath, ".json")}-${timestamp}.json`;
			const backupPath = path.join(this.backupDir, backupFileName);

			await fs.copy(this.characterFilePath, backupPath);

			// Clean up old backups
			await this.cleanupOldBackups();

			logger.info({ backupPath }, "Character backup created");
			return backupPath;
		} catch (error) {
			logger.error(
				{ error: error instanceof Error ? error.message : String(error) },
				"Failed to create character backup",
			);
			return null;
		}
	}

	private async cleanupOldBackups(): Promise<void> {
		try {
			const files = await fs.readdir(this.backupDir);
			const backupFiles = files
				.filter((file: string) => file.endsWith(".json"))
				.map((file: string) => ({
					name: file,
					path: path.join(this.backupDir, file),
					stat: fs.statSync(path.join(this.backupDir, file)),
				}))
				.sort(
					(a: { stat: { mtime: Date } }, b: { stat: { mtime: Date } }) =>
						b.stat.mtime.getTime() - a.stat.mtime.getTime(),
				);

			// Keep only the most recent backups
			const filesToDelete = backupFiles.slice(this.maxBackups);
			for (const file of filesToDelete) {
				await fs.unlink(file.path);
			}

			if (filesToDelete.length > 0) {
				logger.info(`Cleaned up ${filesToDelete.length} old backups`);
			}
		} catch (error) {
			logger.error(
				{ error: error instanceof Error ? error.message : String(error) },
				"Error cleaning up old backups",
			);
		}
	}

	validateModification(modification: Record<string, unknown>): {
		valid: boolean;
		errors: string[];
	} {
		const errors: string[] = [];

		try {
			// Schema validation
			CharacterModificationSchema.parse(modification);
		} catch (error) {
			errors.push(`Schema validation failed: ${(error as Error).message}`);
			return { valid: false, errors };
		}

		// Additional validation rules
		for (const [field, validator] of this.validationRules.entries()) {
			if (modification[field] !== undefined) {
				if (!validator(modification[field])) {
					errors.push(`Invalid ${field}: failed validation rules`);
				}
			}
		}

		// Safety checks
		const bioVal = modification.bio;
		if (Array.isArray(bioVal) && bioVal.length > 20) {
			errors.push("Too many bio elements - maximum 20 allowed");
		}

		const topicsVal = modification.topics;
		if (Array.isArray(topicsVal) && topicsVal.length > 50) {
			errors.push("Too many topics - maximum 50 allowed");
		}

		return { valid: errors.length === 0, errors };
	}

	async applyModification(
		modification: CharacterModification,
	): Promise<{ success: boolean; error?: string }> {
		// Validate modification
		const validation = this.validateModification(modification);
		if (!validation.valid) {
			return {
				success: false,
				error: `Validation failed: ${validation.errors.join(", ")}`,
			};
		}

		try {
			// Create backup first
			await this.createBackup();

			// Get current character
			const currentCharacter = { ...this.runtime.character };
			const previousName =
				typeof this.runtime.character.name === "string"
					? this.runtime.character.name
					: undefined;

			// Apply modifications using merge logic (additive, not replacement)

			// Handle name modification - direct replacement
			if (modification.name) {
				const oldName = currentCharacter.name;
				currentCharacter.name = modification.name;
				logger.info(
					{ oldName, newName: modification.name },
					"Character name changed",
				);
			}

			// Handle system prompt modification - this is a direct replacement, not additive
			if (modification.system) {
				const oldSystem = currentCharacter.system || "No system prompt";
				currentCharacter.system = modification.system;

				logger.info(
					{
						oldLength: oldSystem.length,
						newLength: modification.system.length,
						changed: oldSystem !== modification.system,
					},
					"System prompt modified",
				);
			}

			if (modification.bio) {
				const currentBio = Array.isArray(currentCharacter.bio)
					? currentCharacter.bio
					: typeof currentCharacter.bio === "string"
						? [currentCharacter.bio]
						: [];

				// Add new bio elements, avoiding duplicates
				const newBioElements = modification.bio.filter(
					(newBio) =>
						!currentBio.some(
							(existing) =>
								existing.toLowerCase().includes(newBio.toLowerCase()) ||
								newBio.toLowerCase().includes(existing.toLowerCase()),
						),
				);

				currentCharacter.bio = [...currentBio, ...newBioElements];
			}

			if (modification.topics) {
				const currentTopics = currentCharacter.topics || [];
				const newTopics = modification.topics.filter(
					(topic) => !currentTopics.includes(topic),
				);
				currentCharacter.topics = [...currentTopics, ...newTopics];
			}

			if (modification.messageExamples) {
				const currentExamples = currentCharacter.messageExamples || [];
				const newExampleGroups: MessageExampleGroup[] =
					modification.messageExamples.map((group) => ({
						examples: group.map(
							(item): MessageExample => ({
								name: item.name ?? "",
								content: {
									text: item.content?.text ?? "",
									...(item.content?.actions
										? { actions: item.content.actions }
										: {}),
								},
							}),
						),
					}));
				currentCharacter.messageExamples = [
					...currentExamples,
					...newExampleGroups,
				];
			}

			if (modification.style) {
				currentCharacter.style = {
					...currentCharacter.style,
					...modification.style,
				} as typeof currentCharacter.style;
			}

			if (modification.settings) {
				currentCharacter.settings = {
					...currentCharacter.settings,
					...modification.settings,
				} as typeof currentCharacter.settings;
			}

			// Write to file if available
			if (this.characterFilePath) {
				await fs.writeJSON(this.characterFilePath, currentCharacter, {
					spaces: 2,
				});
				logger.info("Character file updated successfully");
			}

			const persistenceService = this.runtime.getService(
				"character_persistence",
			) as unknown as CharacterPersistenceServiceLike | null;
			if (persistenceService) {
				const persistenceResult = await persistenceService.persistCharacter({
					character: currentCharacter as Record<string, unknown>,
					previousName,
				});
				if (!persistenceResult.success) {
					return {
						success: false,
						error:
							persistenceResult.error ?? "Failed to persist character changes",
					};
				}
			}

			// Update runtime character only after persistence succeeds.
			Object.assign(this.runtime.character, currentCharacter);

			// Audit logging is best-effort; a logging failure must not negate
			// an already-applied character update.
			try {
				const dummyRoomId = this.runtime.agentId; // Use agentId as fallback roomId
				await this.runtime.createMemory(
					{
						entityId: this.runtime.agentId,
						roomId: dummyRoomId,
						content: {
							text: `Character modification applied to file: ${this.characterFilePath || "memory-only"}`,
							source: "character_modification",
						},
						metadata: {
							type: MemoryType.CUSTOM,
							service: PersonalityServiceType,
							action: "character_modified",
							timestamp: Date.now(),
							filePath: this.characterFilePath,
							modificationType: "file_update",
						},
					},
					"character_modifications",
				);
			} catch (error) {
				logger.warn(
					{ error: error instanceof Error ? error.message : String(error) },
					"Character modification audit log failed",
				);
			}

			return { success: true };
		} catch (error) {
			logger.error(
				{ error: error instanceof Error ? error.message : String(error) },
				"Failed to apply character modification",
			);
			return {
				success: false,
				error: `Application failed: ${(error as Error).message}`,
			};
		}
	}

	async getModificationHistory(limit = 10): Promise<
		Array<{
			timestamp: number | undefined;
			modification: unknown;
			filePath: string | undefined;
		}>
	> {
		const memories = await this.runtime.getMemories({
			entityId: this.runtime.agentId,
			count: limit,
			tableName: "character_modifications",
		});

		return memories.map((memory) => {
			const meta = memory.metadata as Record<string, unknown> | undefined;
			return {
				timestamp:
					typeof meta?.timestamp === "number" ? meta.timestamp : undefined,
				modification: meta?.modification,
				filePath:
					typeof meta?.filePath === "string" ? meta.filePath : undefined,
			};
		});
	}

	async getAvailableBackups(): Promise<
		Array<{ path: string; timestamp: number; size: number }>
	> {
		if (!(await fs.pathExists(this.backupDir))) {
			return [];
		}

		const files = await fs.readdir(this.backupDir);
		const backups: Array<{ path: string; timestamp: number; size: number }> = [];

		for (const file of files) {
			if (file.endsWith(".json")) {
				const filePath = path.join(this.backupDir, file);
				const stat = await fs.stat(filePath);

				// Extract timestamp from filename (format: character-YYYYMMDD-HHMMSS.json)
				const timestampMatch = file.match(/character-(\d{8})-(\d{6})\.json/);
				let timestamp = stat.mtime.getTime();

				if (timestampMatch) {
					const dateStr = timestampMatch[1];
					const timeStr = timestampMatch[2];
					const year = parseInt(dateStr.substring(0, 4), 10);
					const month = parseInt(dateStr.substring(4, 6), 10) - 1;
					const day = parseInt(dateStr.substring(6, 8), 10);
					const hour = parseInt(timeStr.substring(0, 2), 10);
					const minute = parseInt(timeStr.substring(2, 4), 10);
					const second = parseInt(timeStr.substring(4, 6), 10);

					timestamp = new Date(
						year,
						month,
						day,
						hour,
						minute,
						second,
					).getTime();
				}

				backups.push({
					path: filePath,
					timestamp,
					size: stat.size,
				});
			}
		}

		return backups.sort((a, b) => b.timestamp - a.timestamp);
	}

	async restoreFromBackup(
		backupPath: string,
	): Promise<{ success: boolean; error?: string }> {
		try {
			// Validate backup file exists and is readable
			if (!(await fs.pathExists(backupPath))) {
				return { success: false, error: "Backup file not found" };
			}

			// Read and validate backup content
			const backupContent = await fs.readJSON(backupPath);

			if (!backupContent.name || typeof backupContent.name !== "string") {
				return {
					success: false,
					error: "Invalid backup file format - missing character name",
				};
			}

			// Create a backup of the current state before restoration
			const currentBackupPath = await this.createBackup();

			// Update runtime character
			Object.assign(this.runtime.character, backupContent);

			// If we have a character file path, update the file
			if (this.characterFilePath) {
				await fs.writeJSON(this.characterFilePath, backupContent, {
					spaces: 2,
				});
			}

			// Log the restoration
			const dummyRoomId = this.runtime.agentId;
			await this.runtime.createMemory(
				{
					entityId: this.runtime.agentId,
					roomId: dummyRoomId,
					content: {
						text: `Character restored from backup: ${path.basename(backupPath)}`,
						source: "character_restoration",
					},
					metadata: {
						type: MemoryType.CUSTOM,
						service: PersonalityServiceType,
						action: "character_reset",
						timestamp: Date.now(),
						backupPath,
						previousBackup: currentBackupPath,
						restorationType: "backup_restoration",
					},
				},
				"character_modifications",
			);

			logger.info(
				{
					backupPath,
					characterName: backupContent.name,
					currentBackup: currentBackupPath,
				},
				"Character restored from backup",
			);

			return { success: true };
		} catch (error) {
			logger.error(
				{ error: error instanceof Error ? error.message : String(error) },
				"Failed to restore from backup",
			);
			return {
				success: false,
				error: `Restoration failed: ${(error as Error).message}`,
			};
		}
	}

	async restoreFromHistory(
		entryIndex: number,
	): Promise<{ success: boolean; error?: string }> {
		const history = await this.getModificationHistory(50);

		if (entryIndex < 0 || entryIndex >= history.length) {
			return { success: false, error: "Invalid history entry index" };
		}

		const entry = history[entryIndex];
		if (!entry.filePath) {
			return {
				success: false,
				error: "No file path available for this history entry",
			};
		}
		if (entry.timestamp === undefined) {
			return {
				success: false,
				error: "No timestamp available for this history entry",
			};
		}
		const entryTimestamp = entry.timestamp;

		// Find the corresponding backup file
		const backups = await this.getAvailableBackups();
		const backup = backups.find(
			(b) => Math.abs(b.timestamp - entryTimestamp) < 60000, // Within 1 minute
		);

		if (!backup) {
			return { success: false, error: "Corresponding backup file not found" };
		}

		return await this.restoreFromBackup(backup.path);
	}

	async stop(): Promise<void> {
		logger.info("CharacterFileManager stopped");
	}
}
